### ISSUE TITLE
Add Augmentations for Robustness Experiments

### ISSUE THREAD

**Issue by SakshayMahna** [JdeRobot/DeepLearningStudio#64]
The robustness experiments include:

- Translations of the Camera
- Rotations of the Camera

Currently, we use the `albumentations` library to introduce the different augmentations. The appropriate function for this use case would be the `Affine` function.

https://albumentations.ai/docs/api_reference/augmentations/geometric/transforms/

However, the camera rotation operation is a Homography Operation, which does not seem to be available in the library.

**Comment by sergiopaniego** [comment]
I have applied Affine using Albumentations to one of the images in the dataset, so we can discuss if it enough with this function or we need to implement something more complex.

## Normal image
![normal](https://user-images.githubusercontent.com/17179696/175508594-3014b84d-ad4a-4c14-aa1b-580911ce815b.png)
## Normal image with affine (Rotation and translation)
![normal_affine](https://user-images.githubusercontent.com/17179696/175508597-36eb9c60-f358-45d7-b07f-0a671d31c439.png)
## Cropped image
![cropped](https://user-images.githubusercontent.com/17179696/175508601-f3acf371-6565-44b2-9385-638f7d909286.png)
## Cropped image with affine (rotation and translation)
![cropped_affine](https://user-images.githubusercontent.com/17179696/175508605-826a4b76-09af-4b7f-a10c-a8e9ff65b70d.png)

What do you think? Could you share image examples of applying the transformation on a model level so we can compare and decide? 

The code that I have used is quite simple: 
```
# Examples made with x=-0.1, x=0.1, y=-0.1, y=0.1
augmentation_option = Compose([Affine(translate_percent={'x': -0.1, 'y': 0}, always_apply=True)])
```

**Comment by SakshayMahna** [comment]
That's a nice trick. Here are the original model images:

## Translation
### Original
![translate](https://user-images.githubusercontent.com/40422728/175545048-f801d976-659d-4978-a4a7-f8ae0886e48c.png)

### Cropped
![cropped_translate](https://user-images.githubusercontent.com/40422728/175545068-acc69e35-7abd-41fc-8a8a-a940696d7f4a.png)

## Rotation
### Original
![rotate](https://user-images.githubusercontent.com/40422728/175545125-6a6b2414-9f69-466c-a953-f4c2bfffdd85.png)

### Cropped
![cropped_rotate](https://user-images.githubusercontent.com/40422728/175545150-60a12633-e0ac-4853-94c7-be5c7ee46dbf.png)

**Comment by SakshayMahna** [comment]
I think the Affine Transform does approximate the translation and rotation, but in some cases, when we would crop the image, it would capture some black background as well. Any idea how to avoid that?

**Comment by sergiopaniego** [comment]
I'm sharing the results for the different types of available border_modes so we can decide if any fits out needs.
# mode=cv2.BORDER_REPLICATE

![normal (1)](https://user-images.githubusercontent.com/17179696/175563829-a246ecdf-a081-407a-a78f-0b250d7fcc9f.png)
![normal_affine (1)](https://user-images.githubusercontent.com/17179696/175563841-68c0f0a7-d425-41ef-9b8d-f26290c15558.png)
![cropped (1)](https://user-images.githubusercontent.com/17179696/175563848-a08f951a-406f-45a3-93e2-b09dae6cd15a.png)
![cropped_affine (1)](https://user-images.githubusercontent.com/17179696/175563852-03f4fdb3-ef42-4283-beaa-2780797f0b8a.png)

**Comment by sergiopaniego** [comment]
# mode=cv2.BORDER_REFLECT
![normal (2)](https://user-images.githubusercontent.com/17179696/175564412-0752363b-e49e-4ab5-82c3-254fe9df3f0d.png)
![Upload
![Uploading cropped (
![cropped_affine (2)](https://user-images.githubusercontent.com/17179696/175564451-1cfc5658-564b-4e50-915b-4c7dc953aa7f.png)
2).png…]()
ing normal_affine (2).png…]()

**Comment by sergiopaniego** [comment]
#  mode=cv2.BORDER_WRAP
![normal (3)](https://user-images.githubusercontent.com/17179696/175564790-1eeccf11-465e-49b2-8d56-5d0f90a04db0.png)
![cropped_affine (3)](https://user-images.githubusercontent.com/17179696/175564810-df25507d-a7b1-47ad-aed0-c563a4955072.png)
![cropped (3)](https://user-images.githubusercontent.com/17179696/175564832-d1152da1-b5ba-49d3-9b64-d1495866c0b2.png)
![normal_affine (3)](https://user-images.githubusercontent.com/17179696/175564797-b2214150-04e8-445a-9f34-df515bbb03cf.png)

**Comment by sergiopaniego** [comment]
# mode=cv2.BORDER_REFLECT_101
![normal (4)](https://user-images.githubusercontent.com/17179696/175565023-0568a4ac-3726-4851-ba53-26558faf9fed.png)
![normal_affine (4)](https://user-images.githubusercontent.com/17179696/175565050-7fd5fe8f-d6f9-4ed0-9dd6-0f6f6a5550be.png)
![cropped (4)](https://user-images.githubusercontent.com/17179696/175565151-26a22071-44c8-4668-b7b5-77c151c63114.png)
![cropped_affine (4)](https://user-images.githubusercontent.com/17179696/175565159-89c66b46-e511-4618-8d3a-d933245105c7.png)

**Comment by sergiopaniego** [comment]
These are the available options.
IMO, mode=cv2.BORDER_REPLICATE or mode=cv2.BORDER_REFLECT are the best two options. What do you think @SakshayMahna ?

The images are not exactly the same as the ones you've shared. I'm suggesting this solution because of its simplicity of integration and implementation but we can still explore the raw opencv option you suggested.

**Comment by SakshayMahna** [comment]
Yes, these two seem to be the best options. Let's go with REPLICATE. I'll try training the model with the same. In the meantime, study the opencv option as well.

**Comment by SakshayMahna** [comment]
I also did some experiments with the homography operation. Here are the results:

## Original Rotation

![original](https://user-images.githubusercontent.com/40422728/177248921-af80a9ba-b465-41e2-9284-3641b83fadea.png)

## Homography Operation

![rotate](https://user-images.githubusercontent.com/40422728/177248970-eadde919-d65c-457e-ab4d-a7bcb0814b3f.png)

## Homography with Adjusted Background

![rotate_nb](https://user-images.githubusercontent.com/40422728/177249000-3f4c50ac-65b8-4041-95bc-741932465ef4.png)

The accuracy of these operations may be much higher. But, the problem with this approach is that they are extremely hardcoded. We have to input specific points (4 for each image) to transform the image. Due to this, we will only have specific rotation values, like -20, -10, +10, +20.
